📚 Guide: Getting Started with The New Architecture (a.k.a. the Playbook)

cortinico opened the pull request:
This pull request contains the playbook for adopting the New Architecture of React Native (composed by the new NativeModule system **TurboModule** and the new Renderer **Fabric**) to your **Android** and **iOS** libraries and apps.

⚠️ **Please note that this document is still a work in progress and it's constantly evolving** ⚠️

At this stage, we're looking for **feedbacks** on the material of any form. Feel free to leave inline comments and suggestions. Users that are willing to **dry-run** the playbook and provided feedback are also more than welcome.

You can **subscribe** to this PR to receive updates whenever we release a new version of the playbook or whenever we update some of the content.

kelset commented:
I would suggest splitting it in multiple pages, it's way too long for a single page.

cortinico commented:
> I would suggest splitting it in multiple pages, it's way too long for a single page.

Yup we're aware of this. We're looking into splitting it into at least 3/4 different pages as it's really hard to consume as it is at the moment.

mrousavy commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+
+The new architecture makes use of CocoaPods.
+
+#### Add Folly and Other Dependencies
+
+We'll need to ensure Folly is configured properly in any projects that consume your library. With CocoaPods, we can use the `compiler_flags` and `dependency` properties to set it up.
+
+Add these to your `Pod::Spec.new` block:
+
+```ruby
+# folly_version must match the version used in React Native
+# See folly_version in react-native/React/FBReactNativeSpec/FBReactNativeSpec.podspec
+folly_version = '2021.06.28.00-v2'
+folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32'
+
+Pod::Spec.new do |s|
+  # ...
+  s.compiler_flags  = folly_compiler_flags
+
+  s.pod_target_xcconfig    = {
+    "HEADER_SEARCH_PATHS" => "\"$(PODS_ROOT)/boost\""
+  }
+
+  s.dependency "React"
+  s.dependency "React-RCTFabric" # This is for fabric component
+  s.dependency "React-Codegen"
+  s.dependency "RCT-Folly", folly_version
+  s.dependency "RCTRequired"
+  s.dependency "RCTTypeSafety"
+  s.dependency "ReactCommon/turbomodule/core"
+  # ...
+end
+```
````
I think this will break Swift Modules, since a few of those Podspecs (e.g. `ReactCommon/turbomodule/core`) do not expose Swift Modules / Modular Headers. This will prevent people from using Swift.

While working on VisionCamera, I've made a PR to fix this in the react-native core repo: https://github.com/facebook/react-native/pull/31858 - but that was a long time ago, not sure if that still builds.

cortinico replied:
cc @sota000
I'm not really familiar with Swift support. Are you suggesting that we add `"DEFINES_MODULE" => "YES"` to `ReactCommon/turbomodule/core` and the other modules this pod depends on?

mrousavy replied:
Yes I think that is the required flag for this. I think those are the three most important points for a Podspec to allow it being used in Swift modules:
* Have `DEFINES_MODULE` set to `YES`
* Ignore all headers by default using `preserve_paths`, because those might be C++ headers and importing them in Swift will break. This will also prevent nameclashes on building, see https://github.com/firebase/firebase-ios-sdk/issues/4035 for more details on that.
  - For those headers that are plain Objective-C headers and don't contain C++, we can add them to `s.source_files`. This will make them importable by other modules, including Swift.
  - For those headers that are Objective-C++ (or C++) and still need to be exported to be importable by other C++ (non-Swift) modules, we can either create a Subspec specifically for those, or wrap them with `#ifdef __cplusplus` as a last resort?
* `s.requires_arc = true` (I think)

I think there are some improvements on Swift's C++ interoperability (see [discussion here](https://forums.swift.org/t/swift-and-c-interoperability-workgroup-announcement/54998)), but I'm not sure how advanced that is (those efforts started 3 years ago)

raedle commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+  _onSubmit: () => {
+    if (this._ref != null)
+      this._ref.**getViewRef****()**.measure(() => {});
```
@cortinico is this intentional? :)

```
this._ref.getViewRef().measure(() => {});
```

cortinico replied:
> @cortinico is this intentional? :)

Obviously now :/ Sadly, the tool we use internally is not working great exporting markdown and having to deal with codeblocks. I've already fixed some of those problems on the Android samples. Thanks for spotting this. Should be fixed now.

raedle commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+  }
+
+  _onSubmit: () => {
+    this._viewRef.**setNativeProps****(**{
```
Also here, are the asterisks intentional?

cortinico replied:
Nope 👍 Fixed as well.

raedle commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+```javascript
+class MyComponent extends React.Component<Props> {
+  _moveToRegion: (region: Region, duration: number) => {
+   **** **UIManager.dispatchViewManagerCommand(
````
Probably remove asterisks here, too

raedle commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+
+```objc
+RCT_EXPORT_METHOD(moveToRegion:(nonnull NSNumber *)reactTag
+                        region:(`NSDictionary`` ``*`)region
````
Are these backticks intended? ;)

cortinico replied:
Nope 👍 Same as above

cwhenderson20 commented:
Llll

thymikee commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+  render() {
+    const {somePropValue} = this.props;
+    return <View
+       onPress={this._onSubmit}
+       ref={this._captureRef}
+       someProp={somePropValue}
+       style={styles.view} />
```
would be great to run Prettier on the document

cortinico replied:
That's actually already done. There is also a CI step that is checking it.

thymikee commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+
+```javascript
+// MyCustomMapNativeComponent.js
+import codegeNativeCommands from 'codegenNativeCommands';
````
Is this supposed to be a relative path or some package?

cortinico replied:
Yup this should be:

```suggestion
import codegenNativeCommands from 'react-native/Libraries/Utilities/codegenNativeCommands';
```

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+The JavaScript spec defines all APIs that are provided by the native module, along with the types of those constants and functions. You will be using Flow to type the input arguments and outputs of your native module’s methods.
+
+By convention, JavaScript spec files are named `Native<MODULE_NAME>.js` and they export a `TurboModuleRegistry` `Spec` object. The following is a basic JavaScript spec template:
```
>The following is a basic JavaScript spec template:

This says JavaScript but the syntax below is Flow. It reads to me like an unstated assumption that one would want to use Flow. Flow is very unpopular in open source, though I'm not sure about RN libraries. (Do we have some stats on Flow vs TS there?)

If I understand correctly, there's some TS version of this. Would it be good to separate this right away into "If you use TypeScript..." and "If you use Flow.." sections?

sota000 replied:
Great point! I am not sure if we have data about flow vs. TS in RN libraries, but regardless, we are definitely considering TS support as it's used a lot more widely in open source. Currently, this playbook only support flow so that we can share the progress earlier (as opposed to waiting for the TS support to be ready).

cortinico replied:
To follow up on this specific point: I've added a banner in the playbook that gives some insights on future TS support: 

![Screenshot 2022-01-06 at 16 53 05](https://user-images.githubusercontent.com/3001957/148419971-3ef1e6f7-9e29-473d-a957-0838d7aad26e.png)

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+#### Supported Flow Types
+
+You will be using [Flow type annotations](https://flow.org/en/docs/types/) to define your JavaScript spec. Keeping in mind that the goal of defining a JavaScript spec is to ensure the generated native interface code is type safe, the set of supported Flow types will be those that can be mapped one-to-one to a corresponding type on the native platform.
```
>You will be using [Flow type annotations](https://flow.org/en/docs/types/) to define your JavaScript spec.

Similarly, "you will be using Flow" sounds prescriptive if TS version exists. (Does it?)

romanonthego replied:
it does (?) at least if we are taking about `react-native-tscodegen` package/comand;
as far as my tests go it works (eg generates code from typescript interfaces) but I never tested actually build in real app.

romanonthego replied:
also also (and I am really sorry if it sound harsh) prioritising flow documentation over typescript is kinda... strange.

I do understand that flow is still used internally by Facebook (Meta).

But outside, in open source react/react-native ecosystem it's almost exclusively TS - I even struggle to recall any recent packages that used flow.

yazankhateeb replied:
I agree with you @romanonthego, and i also think the priority should for typescript since the vast majority of the community using it

sota000 replied:
Thank you for the feedback! We definitely want to prioritize what's useful for the open source community. Our priority was to be transparent and share an early progress of this playbook. Since Meta is using flow internally, using Flow in this playbook was one of the fastest way to showcase what it could look like to adopt the new architecture so that we can gather early feedback. For example, this feedback is very helpful to gauge how important TS is for the community (though we expected this as well).

cortinico replied:
I've reworded a bit to sound less prescriptive + the TypeScript banner just above should also help clarify what's the current status.

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+You will be using [Flow type annotations](https://flow.org/en/docs/types/) to define your JavaScript spec. Keeping in mind that the goal of defining a JavaScript spec is to ensure the generated native interface code is type safe, the set of supported Flow types will be those that can be mapped one-to-one to a corresponding type on the native platform.
+
+<!-- alex ignore savage -->
```
?

cortinico replied:
I had to add a [alex ignore](https://alexjs.com/) here otherwise I would get the following:

```
docs/new-architecture-playbook.md
  66:36-66:45  warning  `primitive` may be insensitive, use `simple`, `indigenous`, `hunter-gatherer` instead  savage  retext-equality
```

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+> To determine if your library is set up for autolinking, check the CocoaPods output after running `pod install` on an iOS project. If you see "auto linking library name", you are all set to go.
+
+### Preparing your JavaScript codebase for the new React Native Renderer (Fabric)
```
By the time I was in this section, I forgot we're still discussing library code. I read "your JavaScript codebase" as a cue that we're talking about the app now. Can we make this something like "Preparing your library code for ..." so that this context is not lost? It's hard to track with multiple levels of header nesting.

cortinico replied:
You're totally right on this point. We're going to move this part out of the "Library" section and just move it to its own page. This will be done in a subsequent PR.

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+The new renderer also known as Fabric doesn’t use the UIManager so direct calls to UIManager will need to be migrated. Historically, calls to UIManager had some pretty complicated patterns. Fortunately, we’ve created new APIs that are much cleaner. These new APIs are forwards compatible with Fabric so you can migrate your code today and they will work properly when you turn on Fabric!
+
+Fabric will be providing new type safe JS APIs that are much more ergonomic than some of the patterns we've seen in product code today. These APIs require references to the underlying component, no longer using the result of `findNodeHandle`. `findNodeHandle` is used to search the tree for a native component given a class instance. This was breaking the React abstraction model. `findNodeHandle` w ...
```
>won’t be compatible with React concurrent mode once we are ready to roll that out.

We don't refer to it as "mode" anymore. Can we say "React 18" here?

cortinico replied:
Updated 👍

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+While we know that all deprecations are a hassle, this guide is intended to help people update components as smoothly as possible. Here are the steps you need to take to get your JS codebase ready for Fabric
+
+1. Migrating findNodeHandle / getting a HostComponent
+2. Migrating `.measure*(`
```
unclosed paren

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+  }
+}
+
+function MyJSComponent(props) {
```
`MyComponent` and `MyJSComponent` seem like unfortunate names in this example. Is the outer component not a "JS" component? It's hard to remember which one is which. I would suggest renaming to something like `Parent` and `Child` or `Panel` and `Button`, etc.

gaearon commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+}
+```
+
+We can’t convert this call to `this._ref.measure` because `this._ref` is an instance to `MyJSComponent`, which is not a HostComponent and thus does not have a `measure` function.
````
I don't believe this claim is correct. I believe the code _above_ does not work (even in Paper) because you can't have a `ref` to a function component. (That's not related to Fabric and is how React works in general.) The only way this could've worked in the past is if `MyJSComponent` is a class or if `MyJSComponent` is a `forwardRef` wrapper.

Concretely, I would suggest converting the child component to a class for this example to make sense. If it was a function you _already_ couldn't `findNodeHandle` on it anyway so you didn't have this problem.

gaearon commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+}
+```
+
+#### Migrating `.measure(`
````
unclosed paren

gaearon commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+Let’s take a look at an example calling `UIManager.measure`. This code might look something like this
+
+```javascript
+if (this._scrollView == null || viewRef == null) {
````
It is not clear what `this._scrollView` or `viewRef` here are. (Refs? `ref.current` values? if they have the same type, why does only one of them have `Ref` suffix?) It would help to have a slightly more complete example.

cortinico replied:
This was simplified by @sammy-SC here https://github.com/facebook/react-native-website/pull/2935

gaearon commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+└── schema.json
+```
+
+#### Extends the abstract class provided by the generated
````
by the generated ... (code?)

cortinico replied:
Updated with `codegen`

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+You may use the following table as a reference for which types are supported and what they map to in each platform:
+
+| Flow Type                 | Nullable Support? | Android (Java)                       | iOS                                                            | Note                                                                     |
```
this table looks borked

<img width="1189" alt="Screenshot 2021-12-14 at 18 28 49" src="https://user-images.githubusercontent.com/810438/146058192-9fe827d7-cad0-4c31-9ef7-2c1a3e5d7842.png">

cortinico replied:
Fixed 👍 I had to manually escape the `|` inside the table which is not great but yeah ¯\_(ツ)_/¯ https://github.com/facebook/docusaurus/issues/376

gaearon commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+### Use a React Native nightly release
+
+At this time, you must use a React Native nightly release in order to get access to the most up to date changes. Eventually, we will recommend targeting a minimum stable open source release.
```
This guide has a lot of manual steps. I wonder if that's representative of what we can expect the final experience for existing apps to be like once Fabric is stable? If not, we should probably clarify that very prominently.

cortinico replied:
> I wonder if that's representative of what we can expect the final experience for existing apps to be like once Fabric is stable? If not, we should probably clarify that very prominently.

Yup that's a great point. I've added a warning on top of the page + a warning in one of the Android section (as that felt the one where we require more manual steps). We can definitely make it more prominent.

We're also looking into automating some of the steps of the playbook as much as possible. To give an example this PR:
- https://github.com/facebook/react-native/pull/32790
Will allow users to enable all the new architecture machinery by turning a property on/off on Android for apps created with the new template. A similar PR has been landed for iOS as well.

cortinico replied:
Added a banner on the intro page 👍 

![Screenshot 2022-01-06 at 17 30 06](https://user-images.githubusercontent.com/3001957/148424997-a269ba49-6357-41b9-a429-bfed1ff22926.png)

gaearon reviewed (commented):
sorry for coming in unannounced — saw this and a few things jumped out at me! hope some of this will be helpful! exciting times :)

yazankhateeb commented:
Is Typescript supported in writing JavaScript spec as well?

slorber commented:
For curious readers like me, here's a convenient link: https://deploy-preview-2879--react-native.netlify.app/docs/next/new-architecture-playbook

sorianog commented:
I just started learning React Native a couple days ago. Would it be helpful if I provided my candid feedback on this PR or reserve my comments to myself?

janicduplessis commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+                   callback:(RCTResponseSenderBlock)callback)
+{
+  // ...
+}
```
I think it's also needed to add

```objc
- (std::shared_ptr<TurboModule>)getTurboModule:(const ObjCTurboModule::InitParams &)params
{
  return std::make_shared<StringGetterSpecJSI>(params);
}
```

cortinico replied:
@sota000 can you confirm this?

hramos replied:
Hey Janic, I looked into the changes I made to `react-native-fbt` when I was writing the first draft of the playbook, and I do see that I have this method, `getTurboModule:`, implemented in the Fbt native module.

It's been a while since I have worked on this code, but based on the fact that further down below in this guide we do not explicitly implement a lookup method for this module in the `TurboModuleManagerDelegate`, it may be necessary to have this method implementation in place in the module.

I think it's okay to go ahead and include this block in the playbook.

cortinico replied:
Included 👍

janicduplessis commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+```groovy
+includeBuild('../node_modules/react-native-gradle-plugin') {
+    dependencySubstitution {
+        substitute(module("com.facebook.react:react")).using(project(":"))
````
I think a better way to enable source build is to use in the top level `build.gradle`:

```groovy
allprojects {
    resolutionStrategy {
        dependencySubstitution {
            substitute module("com.facebook.react:react-native:+") with project(":ReactAndroid")
        }
    }
    
    ...
}
```

This way if someone uses other third party libs they will also use the source build properly.

janicduplessis replied:
Actually this does something different than what I thought, but might still be nice to add this to better enable source builds.

cortinico replied:
> I think a better way to enable source build is to use in the top level `build.gradle`:

Actually, this is not enabling a Gradle Composite build: https://docs.gradle.org/current/userguide/composite_builds.html.

Given that the Gradle plugin needs to be in a included build, we would prefer to follow the same approach for all the included projects (so you'll have only one place where all the deps substitution rules are declared).

janicduplessis commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+    protected native HybridData initHybrid();
+
+    native boolean canCreateTurboModule(String moduleName);
```
I don't think this method is actually needed. Maybe we can remove it for simplicity.

cortinico replied:
Yup you're right, removed.

janicduplessis commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+As you can see, some methods are `native()` which we will implement in C++ in the following paragraph.
+
+```java
+package com.awesomeproject;
````
Imports could be added here so it is copy pastable:

```
import com.facebook.jni.HybridData;
import com.facebook.proguard.annotations.DoNotStrip;
import com.facebook.react.fabric.ComponentFactory;
import com.facebook.soloader.SoLoader;
```

janicduplessis commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+```java
+package com.awesomeproject;
+
+  @DoNotStrip
````
Indentation is wrong here

janicduplessis commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+
+```cpp
+#include <fbjni/fbjni.h>
+#include "MyApplicationTurboModuleManagerDelegate.h"
````
Also need to add `#include "MyComponentsRegistry.h"`

cortinico replied:
Yup nice catch 👍

janicduplessis commented:
Something that would be nice to add in the guide is migrating to `UIManagerHelper` for dispatching events on Android.

rachelnabors commented:
1. Breaking this into multiple pages in the Guides section is a great idea!
2. "Playbook" will be ambiguous to users searching for terms like "migration guide" or "upgrade" using the site's search bar or Google. I recommend you use these words instead for findablity.

AvGcJ commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+end
+
+def pods()
+// Get config
```
```suggestion
# Get config
```

AvGcJ commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+// Get config
+config = use_native_modules!
+
+// Use env variables to turn it on/off.
```
```suggestion
# Use env variables to turn it on/off.
```

AvGcJ commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+fabric_enabled = ENV['USE_FABRIC']
+use_codegen_discovery= ENV['USE_CODEGEN_DISCOVERY']
+
+// Enabled codegen discovery. This will run the codegen at preinstall time.
```
```suggestion
# Enabled codegen discovery. This will run the codegen at preinstall time.
```

AvGcJ commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+use_codegen_discovery= ENV['USE_CODEGEN_DISCOVERY']
+
+// Enabled codegen discovery. This will run the codegen at preinstall time.
+// Files are generated at {pod installation root}/build/generated/ios/
```
```suggestion
# Files are generated at {pod installation root}/build/generated/ios/
```

AvGcJ commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+  end
+end
+
+// Pass the flag to enable fabric to use_react_native!.
```
```suggestion
# Pass the flag to enable fabric to use_react_native!.
```

troZee commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+As of this writing, the latest stable release is `0.66.3`. Once you have upgraded your project to this version successfully, you may proceed to targeting the `0.0.0-20211102-2008-8fef52035` nightly release. You may target this nightly release the same way you’d target any other version of React Native:
+
+```bash
+yarn add react-native@0.0.0-20211205-2008-583471bc4
````
Question: what is a difference between  `0.0.0-20211205-2008-583471bc4` and `0.0.0-20211215-2008-f16cbe59e`. Could you provide a simple changelog ?

troZee replied:
I noticed this [file](https://github.com/facebook/react-native/blob/main/scripts/generate-artifacts.js) does not exist in `0.0.0-20211205-2008-583471bc4`

thymikee replied:
Would it be simpler to point to `react-native@nightly` tag then?

liuyinglao replied:
> Would it be simpler to point to `react-native@nightly` tag then?

agree, it would be great if we could point to react-native@nightly -- the only concern I can think of is nightly build may capture the latest changes that the playbook is not aware of.

cortinico replied:
As a rule of thumb, we tend to run the playbook with a regular cadence. Once we do a run we pick a version of RN native that we will use for the playbook. 

That's true that `react-native@nightly` would also work, but it would create confusion as a user might use a newer version of RN where maybe a script changed and the changes were not reflected (yet) on the playbook.

Till we pin a specific RN stable version (say 0.67.0 or 0.68.0), we prefer to pin to a specific nightly to reduce the variability when running the playbook.

As a side note: if you're willing to do a run on the latest nightly and point out any new problem, that would be very much appreciated 🙏 

> Question: what is a difference between 0.0.0-20211205-2008-583471bc4 and 0.0.0-20211215-2008-f16cbe59e. Could you provide a simple changelog ?

The easiest way would be to run a git log. As you also pointed out, the `generate-artifacts.js` file was missing.

liuyinglao commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2182 @@
+
+```bash
+// Run pod install with the flags
+USE_FABRIC=1 USE_CODEGEN_DISCOVERY=1 pod install
````
based on the nightly build by Dec 28, the playbook's ios part seems to be a bit out-of-date, e.g. I need to run ```USE_FABRIC=1 USE_CODEGEN_DISCOVERY=1 RCT_NEW_ARCH_ENABLED=1 pod install``` here to generate the codes correctly

cortinico replied:
Yup I believe that's correct. @sota000 can you confirm?

sota000 replied:
It should be overriding USE_CODEGEN_DISCOVERY [here](https://github.com/facebook/react-native/blob/ca5aaa766329055f206e51b2eaefcba4f282b05a/scripts/react_native_pods.rb#L31) when RCT_NEW_ARCH_ENABLED is set to 1. I'll check why it's not generate the codes correctly.

sota000 replied:
Oh! I misread, yes, the instruction will be updated to use:
USE_FABRIC=1 RCT_NEW_ARCH_ENABLED=1 pod install

USE_CODEGEN_DISCOVERY=1 should be automatically set when passing RCT_NEW_ARCH_ENABLED=1.

cortinico replied:
Updated 👍

liuyinglao commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+
+_(Please note that this setup requires you to have the React Gradle Plugin configured in the prerequisite step)._
+
+There are two arguments that are required:
```
three?

cortinico commented:
Hey all,
Thank you very much for the amazing feedback, it was really priceless. I've addressed most of the comments, but there are still some left that requires input from iOS or JS side which I'll look through next week.

Next, we're going to split the playbook into sections for better readability. See
- #2882 

> Is Typescript supported in writing JavaScript spec as well?

On the TS side, we know that this would have been a major deal breaker for the majority of our users. We're working really hard on it and we should be able to release more information on a TypeScript codegen really soon.

Finally, as a small side note. This PR was landed on RN repo some days ago:
- https://github.com/facebook/react-native/pull/32790
This will allow you to have a new app create from the default template which already contains all the new architecture machinery. iOS has similar support as well.

As always, any kind of feedback is more than welcome.
Happy new year 🎆

troZee commented:
Thank y'all for providing this playbook. I was trying to migrate https://github.com/callstack/react-native-pager-view/issues/140 into new fabric architecture. I took some notes, while I was migrating and I would like to share my notes with you. Maybe it will help you with fabric migration: 

Migration component to new architecture (Fabric) only for iOS, this document does not cover the migration of native modules 

Used docs: 

https://github.com/facebook/react-native-website/pull/2879/files
https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files

Prerequisites: 
- Migrate RN app to the latest version (0.67 in my case)
- Make sure, gradle 7.X is used 
- Reanimated does not work with nightly version of react native (no aar found)
- library should be generated by bob-builder (any custom setup can cause some issues)
- This guide and codegen does not support TS libs
- Before you start migration, please read the guide a few times to have a better understanding

Glossary: 
Turbomodule - native module without UI
module - refers to native module without UI
component - refers to native module with UI 


JS/TS 

Run below commands in `example` folder 
`yarn add react-native@0.0.0-20211215-2008-f16cbe59e`
**DONT USE THIS VERSION 0.0.0-20211205-2008-583471bc4**
`yarn add react-native-codegen@0.0.12`
Create a template file (or Spec for a turbomodule) in library `src` dir
```js
import codegenNativeComponent from 'react-native/Libraries/Utilities/codegenNativeComponent';
import type { WithDefault } from 'react-native/Libraries/Types/CodegenTypes';
import type { ViewProps } from 'react-native/Libraries/Components/View/ViewPropTypes';

type NativeProps = $ReadOnly<{|
  ...ViewProps,
  // additional props
|}>;

export default (codegenNativeComponent<NativeProps>('RNCYourFancyLibrary'): HostComponent<NativeProps>);
```
IF LIB IS WRITTEN IN TS, SAVE WITHOUT FORMATTING

To be sure, your template/spec file is correctly written, please run belows commands in example folder 
```
node node_modules/react-native-codegen/lib/cli/combine/combine-js-to-schema-cli.js schema.json ../src
```

```
node node_modules/react-native/scripts/generate-specs-cli.js \
 --platform ios \        
 --schemaPath schema.json \
--outputDir ./codegenFiles \
--libraryName PagerView \
--libraryType components

```
- add codegen config to the package. json file in the root folder (https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R96)
- https://github.com/th3rdwave/react-native-safe-area-context/pull/227#discussion_r773793136 use this package.json workaround 
add codegen config to the example package.json
```
"codegenConfig": {
    "libraries": []
  }
```

- Add babel config (https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-43cd302466ad4b557b05644cc7728b0a96657f2c976853802f474c853c16e89fR4)
- TO CONFIRM (I am not sure, if it  is needed): Add the library in example 
(https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-a593fc14fdef41aaf092f774b92cb74bafeda3e9559f8ae4762374bf4af80c92R22) and remove it in post install (https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-a593fc14fdef41aaf092f774b92cb74bafeda3e9559f8ae4762374bf4af80c92R11)

iOS
- Change the podspec in root folder
- Change the Podfile (https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-4fb44f272217447e6b4b3580916d09d23b9055f6a217ad8b26269cfa6031203aR4)
- run `bundle install` in the example folder
- run `bundle exec pod install`
- this command should generate files in `example/ios/build/generated/ios`
- Change AppDelegate.m extension to .mm (https://github.com/th3rdwave/react-native-safe-area-context/pull/227/files#diff-8cbfe8e3302494c09b6d055cfcf7ebd417ba43ffc8ebf6e4a00197dbd09895cbR1)

cortinico commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2182 @@
+...
+  return <RNTMyNativeView />
+// New way
+import RNTMyNativeViewNativeComponent from 'RNTMyNativeViewNativeComponent';
```
Fixed 👍

jurgiles commented on docs/new-architecture-app-modules-ios.md:
```diff
@@ -0,0 +1,166 @@
+---
+id: new-architecture-app-modules-ios
+title: Enabling TurboMudule on iOS
```
Small typo.

cortinico replied:
What is the typo?

kelset replied:
MUdule

kelset replied:
```suggestion
title: Enabling TurboModule on iOS
```
MUdule

cortinico replied:
Thanks for spotting 👍

kelset commented:
Now that this https://github.com/facebook/react-native-website/pull/2892 has been merged, it might be useful to link the new docs section in a few places here and there across the playbook (ex. in the "Enable Fabric for iOS/Android" you could link /fabric-renderer - https://deploy-preview-2879--react-native.netlify.app/docs/next/fabric-renderer)

gaearon commented on docs/new-architecture-library-intro.md:
```diff
@@ -0,0 +1,501 @@
+
+Much of the migration work requires a HostComponent ref to access certain APIs that are only attached to host components (like View, Text, or ScrollView). HostComponents are the return value of calls to `requireNativeComponent`. `findNodeHandle` tunnels through multiple levels of component hierarchy to find the nearest native component.
+
+As a concrete example, this code uses `findNodeHandle` to tunnel from `MyComponent` through to the `View` rendered by `MyJSComponent`.
```
*(Copy pasting my older comments because they're now on outdated files)*

`MyComponent` and `MyJSComponent` seem like unfortunate names in this example. Is the outer component not a "JS" component? It's hard to remember which one is which. I would suggest renaming to something like `Parent` and `Child` or `Panel` and `Button`, etc.

cortinico replied:
Yup this was still on my table 👍 I'll close your [other comment](https://github.com/facebook/react-native-website/pull/2879#discussion_r768930761) in favour of this one as it's more up-to-date.

cortinico replied:
Reworded to be `ParentComponent` vs `ChildComponent` by @sammy-SC here https://github.com/facebook/react-native-website/pull/2935

gaearon commented on docs/new-architecture-library-intro.md:
```diff
@@ -0,0 +1,501 @@
+  _ref: ?React.ElementRef<typeof MyJSComponent>;
+
+  render() {
+    return <MyJSComponent ref={this._captureRef} onSubmit={this._onSubmit} />
```
As far as I know **this "original code" example does not work at all**. This is because you're trying to get a ref to a function component. Function components don't provide refs by default. So this is a confusing example and should be replaced. To fix it, turn the child component into a class. `findNodeHandle` only ever worked with class instances.

cortinico replied:
Fixed by @sammy-SC here https://github.com/facebook/react-native-website/pull/2935

gaearon reviewed (changes requested):
I have a few concerns from the React side of things. I'm not sure if they're yet to be addressed or accidentally got lost due to reshuffling of files but I want to make sure we treat them as land-blocking.

nomi9995 commented on docs/new-architecture-app-modules-android.md:
```diff
@@ -0,0 +1,415 @@
+                        "NDK_TOOLCHAIN_VERSION=clang",
+                        "GENERATED_SRC_DIR=$buildDir/generated/source",
+                        "PROJECT_BUILD_DIR=$buildDir",
+                        "REACT_ANDROID_DIR=$rootDir/../node_modules/react-native/ReactAndroid",
```
I am using react-native@0.66.4 and For supporting new architecture in my app, I have to set these two variables in `build.gradle` like this

![image](https://user-images.githubusercontent.com/36044436/149882345-11ea65a6-46ae-4f2b-b2aa-ab7bf2b3af1b.png)

```
ANDROID_NDK_VERSION = "21.4.7075529"
ANDROID_NDK_PATH="/Users/apple/Library/Android/sdk/ndk/21.4.7075529"
```

and getting this error

fcntl(): Bad file descriptor
/Users/apple/Library/Android/sdk/ndk/21.4.7075529/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.

Besides this, I also have to copy `react-native/packages/react-native-codegen/android` folder to `../node_modules/react-native-codegen/android`
to make it run properly

cortinico replied:
> I am using react-native@0.66.4

The whole playbook relies on using RN nightlies. RN 0.66.x won't work as some of the infrastructure code is missing.

nomi9995 commented on docs/new-architecture-app-renderer-ios.md:
```diff
@@ -0,0 +1,120 @@
+  presets: ['module:metro-react-native-babel-preset'],
+  plugins: [
+    '@babel/plugin-proposal-class-properties',
+    './node_modules/react-native/packages/babel-plugin-codegen'
```
when we install `react-native`, we don't have a packages folder, so the codegen plugin will be
```
plugins: [
    '@babel/plugin-proposal-class-properties',
    'babel-plugin-codegen'
]
```

cortinico replied:
It should actually be: 

```
plugins: [
    '@babel/plugin-proposal-class-properties',
    '@react-native/babel-plugin-codegen'
]
```

But let me cross check 👍

Simek commented on docs/new-architecture-playbook.md:
````diff
@@ -0,0 +1,2176 @@
+}
+```
+
+For an existing native module, you will likely already have one or more instances of `[RCT_EXPORT_METHOD](https://reactnative.dev/docs/native-modules-ios#export-a-native-method-to-javascript)`. To migrate to the new architecture, you’ll need to make sure the method signature makes use of the structs provided by the codegen output.
````
It looks like this formatting issue has been already fixed:
* https://deploy-preview-2879--react-native.netlify.app/docs/next/new-architecture-library-ios#conform-to-the-protocols-provided-by-the-native-interface-code

nomi9995 commented:
did anyone make compatible `react-native-reanimated` with the nightly release?

tido64 commented on docs/new-architecture-app-renderer-ios.md:
```diff
@@ -0,0 +1,120 @@
+
+## 3. Add Babel Plugins
+
+This will trigger the codegen that will run at the metro building time.
```
Can we make this part of the preset? E.g.
```javascript
module.exports = {
  presets: [
    ['module:metro-react-native-babel-preset', { codegen: true }],
  ],
};
```

nomi9995 replied:
it would be great if it implements in this way.

cortinico commented on docs/new-architecture-playbook.md:
```diff
@@ -0,0 +1,2176 @@
+
+### 1. Enable NDK and the native build
+
+> NOTE: In this iteration of the playbook we’re setting up the project to let you build from source. You might notice an increase in your build time because of this. We’re looking into what would be the preferred approach here so please feel free to share your feedbacks.
```
> by making a snapshot signature

Can you clarify? The increase in the build time relies on the fact that RN is currently consumed as a `.aar` from the `node_modules/react-native/android` folder. With a move from a fully build-from-source setup, this won't be the case anymore.

There are other ways to mitigate the build cost which are going to be published on a separate page I'm working on, but that's a bit outside of the scope of this paragraph.

I'm resolving this for now, feel free to reopen if you want to follow up

cortinico commented on docs/new-architecture-app-renderer-android.md:
```diff
@@ -0,0 +1,449 @@
+                return new FabricJSIModuleProvider(
+                    reactApplicationContext,
+                    componentFactory,
+                    new EmptyReactNativeConfig(),
```
Similarly to your other issue, you need to use RN Nightly versions for this to work. Definitely 0.66.4 is not going to work with this setup.

cortinico commented on docs/new-architecture-app-intro.md:
```diff
@@ -0,0 +1,192 @@
+
+While the AGP version should be updated inside the **top level** `build.gradle` file at the `com.android.tools.build:gradle` dependency line.
+
+If you’re set with it, let’s now install the new Gradle plugin which is distributed through a NPM package called [**`react-native-gradle-plugin`**](https://www.npmjs.com/package/react-native-gradle-plugin). You can do so with:
```
Again that's a error caused by using a older RN version. Closing as this should be resolved by using a RN nightly.

cortinico commented on docs/new-architecture-app-intro.md:
````diff
@@ -0,0 +1,192 @@
+If you’re set with it, let’s now install the new Gradle plugin which is distributed through a NPM package called [**`react-native-gradle-plugin`**](https://www.npmjs.com/package/react-native-gradle-plugin). You can do so with:
+
+```bash
+yarn add react-native-gradle-plugin
````
Same as above 👍 Thanks for reporting though

nomi9995 commented on docs/new-architecture-app-modules-android.md:
````diff
@@ -0,0 +1,415 @@
+}
+```
+
+Finally, we need to create a Makefile inside the `src/main/jni` folder called `Android.mk` with the following content:
````
I am using react-native@0.0.0-20220120-2008-2f67f5d68

https://github.com/nomi9995/react-native-turbomodule-fabric-renderer-boilerpate

I am always getting this error from `android.mk`, is there anything, I am doing wrong?
```
C/C++: fcntl(): Bad file descriptor
C/C++: /Users/apple/Library/Android/sdk/ndk/21.4.7075529/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.
[CXX1405] error when building with ndkBuild using /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk: Build command failed.
Error while executing process /Users/apple/Library/Android/sdk/ndk/21.4.7075529/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk APP_ABI=x86 NDK_ALL_ABIS=x86 NDK_DEBUG=1 NDK_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/intermediates/cxx/Debug/1e4i2923/obj NDK_LIBS_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/intermediates/cxx/Debug/1e4i2923/lib APP_CFLAGS+=-Wall APP_CFLAGS+=-Werror APP_CFLAGS+=-fexceptions APP_CFLAGS+=-frtti APP_CFLAGS+=-DWITH_INSPECTOR=1 APP_CPPFLAGS+=-std=c++17 APP_PLATFORM=android-21 APP_STL=c++_shared NDK_TOOLCHAIN_VERSION=clang GENERATED_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/generated/source PROJECT_BUILD_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build REACT_ANDROID_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/../node_modules/react-native/ReactAndroid REACT_ANDROID_BUILD_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/../node_modules/react-native/ReactAndroid/build APP_SHORT_COMMANDS=false LOCAL_SHORT_COMMANDS=false -B -n}
Android NDK: ERROR:/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/src/main/jni/Android.mk:fb: LOCAL_SRC_FILES points to a missing file    
Android NDK: Check that /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/android/app/build/react-ndk/exported/x86/libfb.so exists  or that its path is correct   

fcntl(): Bad file descriptor
/Users/apple/Library/Android/sdk/ndk/21.4.7075529/build/core/prebuilt-library.mk:45: *** Android NDK: Aborting    .  Stop.

C/C++: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk:28: *** Android NDK: Aborting.    .  Stop.
[CXX1405] error when building with ndkBuild using /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/react/jni/Android.mk: Build command failed.
Error while executing process /Users/apple/Library/Android/sdk/ndk/21.4.7075529/ndk-build with arguments {NDK_PROJECT_PATH=null APP_BUILD_SCRIPT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/react/jni/Android.mk APP_ABI=x86 NDK_ALL_ABIS=x86 NDK_DEBUG=1 APP_PLATFORM=android-21 NDK_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/intermediates/cxx/Debug/681pv6ab/obj NDK_LIBS_OUT=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/intermediates/cxx/Debug/681pv6ab/lib NDK_APPLICATION_MK=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/jni/Application.mk THIRD_PARTY_NDK_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/third-party-ndk REACT_COMMON_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon REACT_GENERATED_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/generated/source REACT_SRC_DIR=/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react -j12 --output-sync=none APP_SHORT_COMMANDS=false LOCAL_SHORT_COMMANDS=false -B -n}
Android NDK: WARNING: Ignoring unknown import directory: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/build/generated/source    
Android NDK: /Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk: Cannot find module with tag 'glog' in import path    
Android NDK: Are you sure your NDK_MODULE_PATH variable is properly defined ?    
Android NDK: The following directories were searched:    
Android NDK:         

/Users/apple/Data/Work/reactnative/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native/ReactAndroid/../ReactCommon/butter/Android.mk:28: *** Android NDK: Aborting.    .  Stop.
```

cortinico replied:
> I am always getting this error from `android.mk`, is there anything, I am doing wrong?

Thanks for reporting this. This is definitely a bug on Google's end for AGP: https://issuetracker.google.com/issues/207403732

For now, you should be good to go with the following in your `android/app/build.gradle`:

```diff
afterEvaluate {
    preBuild.dependsOn(packageReactNdkLibs)
+   configureNdkBuildDebug.dependsOn(preBuild)
+   configureNdkBuildRelease.dependsOn(preBuild)
}
```

I've tested it on your project with `yarn && cd android && ./gradlew assembleDebug` and it works on my end.

I will also update the playbook to reflect this.

nomi9995 replied:
@cortinico after resolving the above issue, I am getting another issue
- `codegen` is not creating `com.example.samplelibrary.MyApplicationLibrarySpec` on gradle sync

![image](https://user-images.githubusercontent.com/36044436/150548680-13f88d98-f351-4475-b6cd-09453b918729.png)

- if i delete `android/app/build/react-ndk` than `react-ndk` is not auto creating if i do gradle sync, rebuild project etc

cortinico replied:
> on gradle sync

Nothing of the codegen gets built with a Gradle Sync. We haven't (yet) looked into how to plug this step inside the IDE/Gradle Sync phase. If you invoke `assembleDebug` from the command line OR invoke `yarn react-native run-android` everything should work.

nomi9995 replied:
@cortinico one again, thanks, if codegen on Gradle sync is required then I can contribute this one because I have already done this type of task in my open source project 
https://github.com/nomi9995/react-native-keys/blob/master/android/RNKeys.gradle

can you confirm for me if this codegen in Gradle sync is required in the future or not

nomi9995 replied:
@cortinico another issue that I am getting when I try to run `./gradlew assemble`
```
/Users/apple/Data/Work/shobbak/reactnative/packages/ReactNativeTurbomoduleFabricRendererBoilerpate/node_modules/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/ReactExtension.kt: (10, 37): 'BaseVariant' is deprecated. Deprecated in Java
```
![image](https://user-images.githubusercontent.com/36044436/150551975-e6d233f7-dd9c-49b7-8768-4f49dad1b0e0.png)

cortinico replied:
> can you confirm for me if this codegen in Gradle sync is required in the future or not

We haven't discussed on this front yet (cc @ShikaSD). Thanks for your input though 👍 

> another issue that I am getting when I try to run ./gradlew assemble

That's a build warning that we had as we moved to AGP 7.x as the `'BaseVariant'` is deprecated and we should move to use the newer variant API here. Not a blocker for this guide though.

nomi9995 replied:
@cortinico I have updated react-native nightly to latest nightly release but i am unable to run the android app because - codegen is not creating files
- ios app is running but i am unable to call `getString` function in `App.tsx`
is there i am doing something wrong?

https://github.com/nomi9995/react-native-turbomodule-fabric-renderer-boilerpate

![image](https://user-images.githubusercontent.com/36044436/150652787-fb1f0971-0a06-4deb-8a8c-d7db415a0c0d.png)

nomi9995 replied:
@cortinico ?

cortinico commented on docs/new-architecture-app-intro.md:
````diff
@@ -0,0 +1,192 @@
+Now, you can edit your **top level** `settings.gradle` file to include the following line at the end of the file:
+
+```groovy
+includeBuild('../node_modules/react-native-gradle-plugin') {
````
> I pulled branch`0.66-stable` and copied android folder

That's not the correct approach. There should not be an `android` folder inside `react-native-codegen` therefore this can't be the long term solution. I'm trying to see if I can reproduce this error locally, but it seems like this is not the case.

cortinico commented on docs/new-architecture-app-renderer-android.md:
```diff
@@ -0,0 +1,449 @@
+
+    @Override
+    protected ReactRootView createRootView() {
+      ReactRootView reactRootView = new ReactRootView(getContext());
```
> it should be MainActivity.this like this.

That's essentially equivalent. Any specific reason why it should be preferred?

ShikaSD commented:
> Something that would be nice to add in the guide is migrating to `UIManagerHelper` for dispatching events on Android.

@janicduplessis what kind of migration do you mean? Is it just using `UIManagerHelper` instead of accessing `UIManager` or `EventDispatcher` directly? Would be great to have code example of what you had in mind :)

janicduplessis commented:
> > Something that would be nice to add in the guide is migrating to `UIManagerHelper` for dispatching events on Android.
> 
> @janicduplessis what kind of migration do you mean? Is it just using `UIManagerHelper` instead of accessing `UIManager` or `EventDispatcher` directly? Would be great to have code example of what you had in mind :)

Yea, in my case it was for dispatching events.

I had something like this before

```java
EventDispatcher dispatcher =
  reactContext.getNativeModule(UIManagerModule.class).getEventDispatcher();
dispatcher.dispatchEvent(new InsetsChangeEvent(view.getId(), insets, frame));
```

and after

```java
int reactTag = view.getId();
UIManagerHelper.getEventDispatcherForReactTag(reactContext, reactTag)
  .dispatchEvent(
    new InsetsChangeEvent(
      UIManagerHelper.getSurfaceId(reactContext), reactTag, insets, frame));
```

cortinico commented on docs/new-architecture-app-intro.md:
````diff
@@ -0,0 +1,192 @@
+Now you will have your AppDelegate conform to `RCTCxxBridgeDelegate`. Start by adding the following imports at the top of your AppDelegate file:
+
+```objc
+#import <React/HermesExecutorFactory.h>
````
You're right 👍  Updated.

nomi9995 commented on docs/new-architecture-app-modules-android.md:
```diff
@@ -0,0 +1,425 @@
+
+import java.util.List;
+
+public class MyApplicationTurboModuleManagerDelegate extends ReactPackageTurboModuleManagerDelegate {
```
Ios is working well. but when I run the android app then I am getting this error. it there I am doing something wrong. I just followed the documentation.

```
E/log: error java.lang.ClassNotFoundException: Didn't find class "com.reactnative_turbomodule_fabricrenderer_boilerpate.MyApplicationTurboModuleManagerDelegate" on path: DexPathList[[zip file "/data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/base.apk"],nativeLibraryDirectories=[/data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/lib/x86_64, /data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/base.apk!/lib/x86_64, /system/lib64]]
2022-01-25 15:40:58.754 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542] JNI DETECTED ERROR IN APPLICATION: JNI NewGlobalRef called with pending exception java.lang.ClassNotFoundException: Didn't find class "com.reactnative_turbomodule_fabricrenderer_boilerpate.MyApplicationTurboModuleManagerDelegate" on path: DexPathList[[zip file "/data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/base.apk"],nativeLibraryDirectories=[/data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/lib/x86_64, /data/app/com.reactnativeturbomodulefabricrendererboilerpate-hPxyPOnGJ_lkF_YfVWm8OA==/base.apk!/lib/x86_64, /system/lib64]]
2022-01-25 15:40:58.755 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542] (Throwable with no stack trace)
2022-01-25 15:40:58.755 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542] 
2022-01-25 15:40:58.755 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542]     in call to NewGlobalRef
2022-01-25 15:40:58.755 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542]     from java.lang.String java.lang.Runtime.nativeLoad(java.lang.String, java.lang.ClassLoader)
2022-01-25 15:40:58.755 30666-30716/com.reactnativeturbomodulefabricrendererboilerpate A/dererboilerpat: java_vm_ext.cc:542] "create_react_context" prio=5 tid=30 Runnable
```
repro:
https://github.com/nomi9995/react-native-turbomodule-fabric-renderer-boilerpate

nomi9995 replied:
this was resolved

nomi9995 commented on docs/new-architecture-app-modules-android.md:
```diff
@@ -0,0 +1,425 @@
+    }
+
+    protected native HybridData initHybrid();
+
```
here we need to add this line
`native boolean canCreateTurboModule(String moduleName);`
otherwise, I am getting this error in logcat
```
W/dererboilerpat: CheckJNI: method to register "canCreateTurboModule" not in the given class. This is slow, consider changing your RegisterNatives calls.
E/dererboilerpat: ----- class 'Lcom/reactnativeturbomodulefabricrendererboilerpate/MyApplicationTurboModuleManagerDelegate;' cl=0x13cc2420 -----
E/dererboilerpat:   objectSize=265 (120 from super)
E/dererboilerpat:   access=0x0008.0001
E/dererboilerpat:   super='java.lang.Class<com.facebook.react.ReactPackageTurboModuleManagerDelegate>' (cl=0x13cc2420)
E/dererboilerpat:   vtable (2 entries, 3 in super):
E/dererboilerpat:      0: com.facebook.jni.HybridData com.reactnativeturbomodulefabricrendererboilerpate.MyApplicationTurboModuleManagerDelegate.initHybrid()
E/dererboilerpat:      1: void com.reactnativeturbomodulefabricrendererboilerpate.MyApplicationTurboModuleManagerDelegate.maybeLoadOtherSoLibraries()
E/dererboilerpat:   direct methods (1 entries):
E/dererboilerpat:      0: void com.reactnativeturbomodulefabricrendererboilerpate.MyApplicationTurboModuleManagerDelegate.<init>(com.facebook.react.bridge.ReactApplicationContext, java.util.List)
E/dererboilerpat:   static fields (1 entries):
E/dererboilerpat:      0: boolean com.reactnativeturbomodulefabricrendererboilerpate.MyApplicationTurboModuleManagerDelegate.sIsSoLibraryLoaded
E/dererboilerpat: 
E/dererboilerpat: Failed to register native method com.reactnativeturbomodulefabricrendererboilerpate.MyApplicationTurboModuleManagerDelegate.canCreateTurboModule(Ljava/lang/String;)Z in /data/app/com.reactnativeturbomodulefabricrendererboilerpate-P3yaQha3eGo-5UIRn2clnw==/base.apk!classes19.dex
E/log: error java.lang.NoSuchMethodError: no static or non-static method "Lcom/reactnativeturbomodulefabricrendererboilerpate/MyApplicationTurboModuleManagerDelegate;.canCreateTurboModule(Ljava/lang/String;)Z"
A/dererboilerpat: java_vm_ext.cc:542] JNI DETECTED ERROR IN APPLICATION: JNI NewGlobalRef called with pending exception java.lang.NoSuchMethodError: no static or non-static method "Lcom/reactnativeturbomodulefabricrendererboilerpate/MyApplicationTurboModuleManagerDelegate;.canCreateTurboModule(Ljava/lang/String;)Z"
A/dererboilerpat: java_vm_ext.cc:542] (Throwable with no stack trace)
A/dererboilerpat: java_vm_ext.cc:542] 
A/dererboilerpat: java_vm_ext.cc:542]     in call to NewGlobalRef
A/dererboilerpat: java_vm_ext.cc:542]     from java.lang.String java.lang.Runtime.nativeLoad(java.lang.String, java.lang.ClassLoader)
A/dererboilerpat: java_vm_ext.cc:542] "create_react_context" prio=5 tid=28 Runnable
```

cortinico replied:
Thanks for the heads up. Yeah I was investigating this one. @RSNara Can you confirm that `canCreateTurboModule` is a required method for a TM?

cortinico replied:
I've just cross checked once more: `canCreateTurboModule` is not necessary, and you should be good to go building without it.

cortinico commented on docs/new-architecture-app-intro.md:
````diff
@@ -0,0 +1,192 @@
+```objc
+#pragma mark - RCTCxxBridgeDelegate
+
+- (std::unique_ptr<facebook::react::JSExecutorFactory>)jsExecutorFactoryForBridge:(RCTBridge *)bridge
````
Thanks for trying 👍 You're right. I've verified it locally and included them.

Simek commented on docs/new-architecture-app-renderer-ios.md:
```diff
@@ -0,0 +1,128 @@
+
+## 2. Update your root view
+
+The way to render your app with Fabric depends on your setup. Here is an example of how you can enable Fabric in your app with the RN_FABRIC_ENABLED compiler flag to enable/disable. Refer[RN-Tester’s AppDelegate](https://github.com/facebook/react-native/blob/main/packages/rn-tester/RNTester/AppDelegate.mm) as an example.
```
```suggestion
The way to render your app with Fabric depends on your setup. Here is an example of how you can enable Fabric in your app with the `RN_FABRIC_ENABLED` compiler flag to enable/disable. Refer [RN-Tester’s AppDelegate](https://github.com/facebook/react-native/blob/main/packages/rn-tester/RNTester/AppDelegate.mm) as an example.
```
Small formatting tweaks.

Simek commented on docs/new-architecture-app-renderer-ios.md:
````diff
@@ -0,0 +1,128 @@
+  :fabric_enabled => fabric_enabled
+)
+```
+
+```bash
+$ ~/yourproject/**/ios; pod install
+```
````
```suggestion
```

I'm not sure if running `pod install` here is required, since the guide recommends this again as a last step (4.).

If we want to leave this in, it might be worth adding an informative sentence above the command, for example: 
* "Then, update your project pods by running:"

cortinico replied:
Yup the last step should be sufficient. Moreover here there is a missing `cd` so it's sort of confusing for the reader.

cortinico commented:
Hey all,
I wanted to give a short update on this PR.

First I wanted to thank you all on behalf of the whole team for the amazing feedbacks you all left. We're getting closer to merging this first "Getting Started" guide, which will happen sometimes next week. 

If you have further feedback, feel free to leave it asap. As the number of comments on this PR grew to a point where is getting harder to follow every discussion, please consider that we might postpone some follow-ups to separate PRs.

[As mentioned in this blogpost](https://reactnative.dev/blog/2022/01/21/react-native-h2-2021-recap#the-new-architecture-rollout-and-releases), this is going to be the first step to support the OSS community with the New Architecture rollout. We're also looking into ways to let the community be more involved into the feedback loop for the New Architecture rollout, so stay tuned if you wish to share your opinion/idea/concerns.

yaaliuzhipeng commented:
hey,guys . don't you think this is still hard for someone not that familiar with library development ? I think a blog to demonstrate how to create a library with turbomodule enabled  from 0 to 1 is indeed needed 😍

cortinico commented:
> Is Typescript supported in writing JavaScript spec as well?

Following up here as it was a long requested feature. As of today, you can use TypeScript to write your specs for the New Architecture. We've update the docs and opened a thread to discuss it in the working group:

- https://github.com/facebook/react-native-website/pull/3036
- https://github.com/reactwg/react-native-new-architecture/discussions/27
